✨ Support Node Auto Placement and Node AF/AAF

zhanggbj opened the pull request:
<!--  Thanks for sending a pull request! Please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api-provider-vsphere/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other)
Here are some other tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
Support Node Auto Placement and Node AF/AAF

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

zhanggbj commented:
/test pull-cluster-api-provider-vsphere-e2e-govmomi-blocking-main
/test pull-cluster-api-provider-vsphere-e2e-govmomi-conformance-ci-latest-main
/test pull-cluster-api-provider-vsphere-e2e-govmomi-conformance-main
/test pull-cluster-api-provider-vsphere-e2e-govmomi-main
/test pull-cluster-api-provider-vsphere-e2e-govmomi-upgrade-1-34-1-35-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-blocking-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-conformance-ci-latest-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-conformance-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-upgrade-1-34-1-35-main
/test pull-cluster-api-provider-vsphere-e2e-vcsim-govmomi-main
/test pull-cluster-api-provider-vsphere-e2e-vcsim-supervisor-main
/test pull-cluster-api-provider-vsphere-test-main
/test pull-cluster-api-provider-vsphere-verify-main
/test pull-cluster-api-provider-vsphere-apidiff-main
/test pull-cluster-api-provider-vsphere-janitor-main

silvery1622 commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,346 @@
+		// Add per-md-zone label for day-2 operations once placement of a VM belongs to MachineDeployment is done
+		// Do not update per-md-zone label once set, as placement decision should not change without user explicitly
+		// ask.
+		placementDecisionLabels, err := GenerateVMGPlacementAnnotations(ctx, desiredVMG, mdNames)
```
If a node pool is deleted and then recreated with the same name, would the machine of the node pool still use the same placement?

DanielXiao commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,112 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+			if apierrors.IsNotFound(err) {
+				log.V(4).Info("VirtualMachineGroup not found, requeueing", "Name", key.Name, "Namespace", key.Namespace)
+				return true, nil
+			}
```
This is Redundant IsNotFound check

```
	if apierrors.IsNotFound(err) {
		log.V(4).Info("VirtualMachineGroup not found, requeueing", "Name", key.Name, "Namespace", key.Namespace)
		return true, nil
	}
	return false, err
```

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,350 @@
+				}
+
+				log.Info(fmt.Sprintf("VM %s in VMG %s/%s has been placed in zone %s", member.Name, vmg.Namespace, vmg.Name, zone))
+				annotations[fmt.Sprintf("zone.cluster.x-k8s.io/%s", md)] = zone
```
Will fix it in next patch.

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,346 @@
+			}
+
+			// Check if VM belongs to a Machine Deployment by name (e.g. cluster-1-np-1-vm-xxx contains np-1)
+			// TODO: Establish membership via the machine deployment name label
```
Will refine this by checking accurate VM's MachineDeployment label.

silvery1622 commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,350 @@
+
+	var result []vmwarev1.VSphereMachine
+	for _, vs := range vsMachineList.Items {
+		if vs.DeletionTimestamp.IsZero() {
```
Please add log for filtered machines.

silvery1622 commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,350 @@
+			result = append(result, vs)
+		}
+	}
+	log.V(4).Info("Final list of VSphereMachines for VMG member generation", "count", len(result))
```
Why log the list length rather than the list member?

zhanggbj replied:
I think logging hundreds of VM Names could have a cost, especially during we cluster scale/upgrade/rollout/etc., so logging the count for debugging.

silvery1622 commented on main.go:
```diff
@@ -482,6 +486,12 @@ func setupSupervisorControllers(ctx context.Context, controllerCtx *capvcontext.
 		return err
 	}
 
+	if feature.Gates.Enabled(feature.NamespaceScopedZones) && feature.Gates.Enabled(feature.NodeAutoPlacement) {
```
Is a webhook validation needed to deny the mixed mode for creation?

silvery1622 commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,350 @@
+}
+
+// GenerateVMGPlacementAnnotations returns annotations per MachineDeployment which contains zone info for placed VMs for day-2 operations.
+func GenerateVMGPlacementAnnotations(ctx context.Context, vmg *vmoprv1.VirtualMachineGroup, machineDeployments []string) (map[string]string, error) {
```
Update function name to GenerateVirtualMachineGroupPlacementAnnotations as other function use the full name. Eg. AddVirtualMachineGroupControllerToManager

DanielXiao commented on controllers/vmware/virtualmachinegroup_controller.go:
```diff
@@ -0,0 +1,138 @@
+
+	if err != nil {
+		return nil
+	}
```
We'd better add comments to explain the whole logic of the controller. It 's a workaround approach after all.

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,350 @@
+		return reconcile.Result{}, err
+	}
+	mdNames := []string{}
+	for _, md := range machineDeployments.Items {
```
I didn't quite get it. These are two files,  virtualmachinegroup_reconciler.go and vmopmachine.go?

zhanggbj commented on controllers/vmware/virtualmachinegroup_controller.go:
```diff
@@ -0,0 +1,138 @@
+	})
+
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&vmoprv1.VirtualMachineGroup{}).
```
Make sense to me. Will update to watch Cluster as primary, since Cluster triggers VMG lifecycle.

silvery1622 commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,350 @@
+		}
+
+		if expected == 0 {
+			log.Info("none of MachineDeployments specifies replica and node auto replacement doesn't support this scenario")
```
I understand the expect behavior here. Just wondering if we could make some adjustment in the log message.

DanielXiao commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+		}
+
+		if expected == 0 {
+			errMsg := fmt.Sprintf("Found 0 desired Machine for Cluster %s/%s", cluster.Name, cluster.Namespace)
```
Found 0 desired VSphereMachine

zhanggbj replied:
Updated.

sbueringer commented on controllers/vmware/virtualmachinegroup_controller.go:
```diff
@@ -0,0 +1,109 @@
+		Named("virtualmachinegroup").
+		Watches(
+			&vmoprv1.VirtualMachineGroup{},
+			handler.EnqueueRequestForOwner(mgr.GetScheme(), reconciler.Client.RESTMapper(), &clusterv1.Cluster{}),
```
```suggestion
			handler.EnqueueRequestForOwner(mgr.GetScheme(), reconciler.Client.RESTMapper(), &clusterv1.Cluster{}),
			ctrlbldr.WithPredicates(predicates.ResourceIsChanged(mgr.GetScheme(), predicateLog)),
```

We have not been using this in CAPV up until now, but let's use it here to already improve this new reconciler (I have a task to follow-up for others soon'ish).

Per default the reconciler will get resyncs from all informers. i.e. every 10m everything gets reconciled because of the Cluster resync and additionally every 10m everything gets reconciled because of the VirtualMachineGroup resync.

In core CAPI we drop events from resyncs of secondary watches to avoid this issue

(it's not needed for VSphereMachine below as we already drop all updates)

zhanggbj replied:
Good catch. This help to reduce the resync noices.

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+	// reconciliationDelay is the delay time for requeueAfter.
+	reconciliationDelay = 10 * time.Second
+	// ZoneAnnotationPrefix is the prefix used for placement decision annotations which will be set on VirtualMachineGroup.
+	ZoneAnnotationPrefix = "zone.cluster.x-k8s.io"
```
I think we have to namespace this to CAPV as it's not a Cluster API wide standard annotation

(TBD with @fabriziopandini what exactly we should use instead)

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+	Recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch
```
Let's keep these here and drop the duplicates in virtualmachinegroup_controller.go

zhanggbj replied:
Fixed.

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+	}
+
+	// If ControlPlane haven't initialized, requeue it since VSphereMachines of MachineDeployment will only be created after
+	// ControlPlane is initialized.
```
I agree that we should wait, I am not sure if this statement is true (IIRC it is not)

Could be that VirtualMachine objects will only be created after CP initialized?

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -163,6 +165,13 @@ func (v *VmopMachineService) SyncFailureReason(_ context.Context, machineCtx cap
 	return supervisorMachineCtx.VSphereMachine.Status.FailureReason != nil || supervisorMachineCtx.VSphereMachine.Status.FailureMessage != nil, nil
 }
 
+// affinityInfo is an internal to store VM affinity information.
```
nit
```suggestion
// affinityInfo is an internal struct used to store information about VM affinity.
```

zhanggbj replied:
Fixed.

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+	// Sort the VM names alphabetically for consistent ordering
+	sort.Slice(vmNames, func(i, j int) bool {
+		return vmNames[i] < vmNames[j]
+	})
```
```suggestion
	slices.Sort(vmNames)
```

zhanggbj replied:
Fixed.

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+	if desiredVMG.Labels == nil {
+		desiredVMG.Labels = make(map[string]string)
+		desiredVMG.Labels[clusterv1.ClusterNameLabel] = cluster.Name
+	}
```
```suggestion
	}
	desiredVMG.Labels[clusterv1.ClusterNameLabel] = cluster.Name
```

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+	}
+
+	// Add per-md-zone label for day-2 operations once placement of a VM belongs to MachineDeployment is done.
+	// Do not update per-md-zone label once set, as placement decision should not change without user explicitly
```
Let's please definitely just handle the annotations in the func and make sure we have appropriate unit test coverage.

The way this is done right now the VMG.Annotations map is already modified inside the func anyway (given how pointers / maps work ...)

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+
+		// Skip it if member's VirtualMachineGroupMemberConditionPlacementReady is still not true.
+		if !conditions.IsTrue(&member, vmoprv1.VirtualMachineGroupMemberConditionPlacementReady) {
+			log.Info(fmt.Sprintf("Member %s of %s/%s is not PlacementReady, skipping it", member.Name, vmg.Name, vmg.Namespace))
```
This can also produce a huge amount of logs, right?
I would drop the log entirely

I think object YAMLs will be good enough to debug situations like this

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+			// Get the VM placement information by member status.
+			// VMs that have undergone placement do not have Placement info set, skip.
+			if member.Placement == nil {
+				log.V(4).Info(fmt.Sprintf("VM %s in VMG %s/%s has no placement info. Placement is nil", member.Name, vmg.Name, ns))
```
```suggestion
				log.V(5).Info(fmt.Sprintf("VM %s in VMG has no placement info. Placement is nil", member.Name))
```

the VMG will already be a k/v pair on the logger after the change suggested above

Please also use namespace/name consistently  and not sometimes use name/namespace

Similar for the next log below and the one lin l.375

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+		// Proceed only if the machine is a member of the VirtualMachineGroup.
+		isMember, err := v.checkVirtualMachineGroupMembership(vmOperatorVMGroup, supervisorMachineCtx)
+		if err != nil {
+			return true, errors.Wrapf(err, "%s", fmt.Sprintf("failed to check if VirtualMachine %s is a member of VirtualMachineGroup %s/%s", supervisorMachineCtx.VSphereMachine.Name, vmOperatorVMGroup.Name, vmOperatorVMGroup.Namespace))
```
Please use klog.KObj instead of %s/%s for secondary objects.
```suggestion
			return true, errors.Wrapf(err, "%s", fmt.Sprintf("failed to check if VirtualMachine %s is a member of VirtualMachineGroup %s", supervisorMachineCtx.VSphereMachine.Name, klog.KObj(vmOperatorVMGroup)))
```

Also please do not add the name of the primary object on errors (they are already linked to the primary object)

zhanggbj replied:
Fixed.

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+		}
+
+		// Set the zone label using the annotation of the per-md zone mapping from VMG.
+		// This is for new VMs created during day-2 operations in VC 9.1.
```
Please fix this comment as discussed

sbueringer commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+
+		// Set the zone label using the annotation of the per-md zone mapping from VMG.
+		// This is for new VMs created during day-2 operations in VC 9.1.
+		nodePool := supervisorMachineCtx.Machine.Labels[clusterv1.MachineDeploymentNameLabel]
```
```suggestion
		mdName := supervisorMachineCtx.Machine.Labels[clusterv1.MachineDeploymentNameLabel]
```

Please check if there are more places where we added it in this PR. There are no nodePools in CAPI

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+		expected, err := getExpectedVSphereMachineCount(ctx, r.Client, cluster)
+		if err != nil {
+			log.Error(err, "failed to get expected Machines of all MachineDeployment")
+			return reconcile.Result{}, err
```
Sorry this comment belongs to `if expected == 0 {` in l.142

sbueringer commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+			continue
+		}
+
+		// Get VSphereMachine which share the same Name of the member Name and get the MachineDeployment Name it belonged to.
```
Discussed with Fabrizio, all good

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+	var affInfo affinityInfo
+	if feature.Gates.Enabled(feature.NodeAutoPlacement) &&
+		!infrautilv1.IsControlPlaneMachine(machineCtx.GetVSphereMachine()) {
+		vmOperatorVMGroup := &vmoprv1.VirtualMachineGroup{}
```
may be
```suggestion
		vmGroup := &vmoprv1.VirtualMachineGroup{}
```
(it goes without saying it is from VM operator, so let's get rid of it to improve readability)

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+				},
+			},
+			VMAntiAffinity: &vmoprv1.VMAntiAffinitySpec{
+				PreferredDuringSchedulingPreferredDuringExecution: []vmoprv1.VMAffinityTerm{
```
```suggestion
                 // All the machines belonging to the same MachineDeployment should be spread across esxi hosts in the same failure domain
				PreferredDuringSchedulingPreferredDuringExecution: []vmoprv1.VMAffinityTerm{
```

fabriziopandini replied:
This comment seems not yet addressed

fabriziopandini commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+//
+// - Batch Coordination: Gating the initial creation of the VMG until all expected worker
+// VSphereMachines are present. This ensures the complete VM member list is sent to the VM
+// Service in a single batch operation due to a limitation of underlying service.
```
```suggestion
// VSphereMachines are present.
```
I would drop this sentence and clarify a little bit what is the definition of all the expected worker.

fabriziopandini commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+		}
+
+		// Wait for all intended VSphereMachines corresponding to MachineDeployment to exist only during initial Cluster creation.
+		current := int32(len(current))
```
this is very confusing, we are shadowing a variable that contains a list of virtual machines with another with the same name, but different type and meaning.


```suggestion
		currentVSphereMachineCount := int32(len(current))
```

zhanggbj replied:
Fixed.

fabriziopandini commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+		// Calculate expected Machines of all MachineDeployments.
+		// CAPV retrieves placement decisions from the VirtualMachineGroup to guide
+		// day-2 VM placement. At least one VM is required for each MachineDeployment.
+		expected, err := getExpectedVSphereMachineCount(ctx, r.Client, cluster)
```
should this be

```suggestion
		expectedVSphereMachineCount, err := getExpectedVSphereMachineCount(ctx, r.Client, cluster)
```

zhanggbj replied:
Fixed.

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+			affInfo.failureDomain = ptr.To(zone)
+		}
+
+		// Fetch machine deployments without explicit failureDomain specified
```
I'm wondering if we should skip computing AffinitySpec after the machine has been created (we not using it for update)

zhanggbj replied:
AffinitySpec doesn't impact after the initial cluster creation, but to keep it consistent, I think we should keep computing here. Especially in future, if AffinitySpec start to take effect during day-2, we don't need to update the code here.

zhanggbj commented on controllers/vmware/virtualmachinegroup_controller.go:
```diff
@@ -0,0 +1,109 @@
+
+	if err != nil {
+		return nil
+	}
```
Make senses. Fixed it and also added predict in watch for VSphereMachine to only handle VSphereMachines which have MachineDeployment label.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+				return false, err
+			}
+			if apierrors.IsNotFound(err) {
+				log.V(4).Info("VirtualMachineGroup not found, requeueing", "Name", key.Name, "Namespace", key.Namespace)
```
Fixed.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+				Status: metav1.ConditionFalse,
+				Reason: infrav1.VSphereMachineVirtualMachineWaitingForVirtualMachineGroupV1Beta2Reason,
+			})
+			log.V(4).Info("Waiting for VirtualMachineGroup membership, requeueing", "VM Name", supervisorMachineCtx.Machine.Name)
```
Fixed.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+								clusterv1.MachineDeploymentNameLabel: nodePool,
+							},
+						},
+						TopologyKey: corev1.LabelTopologyZone,
```
VMOP is using the corev1 Labels Key, such as https://github.com/search?q=repo%3Avmware-tanzu%2Fvm-operator%20corev1.LabelTopologyZone&type=code. Seems their affinity design is following k8s pod affinity design.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+				Status: metav1.ConditionFalse,
+				Reason: infrav1.VSphereMachineVirtualMachineWaitingForVirtualMachineGroupV1Beta2Reason,
+			})
+			log.V(4).Info("Waiting for VirtualMachineGroup membership, requeueing", "VM Name", supervisorMachineCtx.Machine.Name)
```
Fixed.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -494,6 +609,15 @@ func (v *VmopMachineService) reconcileVMOperatorVM(ctx context.Context, supervis
 			vmOperatorVM = typedModified
 		}
 
+		if affinityInfo != nil && affinityInfo.affinitySpec != nil {
```
Added comments for clarification that affinity only impact VM locations during day-1.

> I would also probably revert the if, to check "if not set" (set on creation) first

Could you help clarify this one?

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -163,6 +165,13 @@ func (v *VmopMachineService) SyncFailureReason(_ context.Context, machineCtx cap
 
+// affinityInfo is an internal to store VM affinity information.
+type affinityInfo struct {
+	affinitySpec  *vmoprv1.AffinitySpec
```
Oh right, whenever we set affinityInfo, we'll always set affinitySpec. Seems only checking affinityInfo != nil is fine.

However, since AffinitySpec contains lots of rules and arrays of MachineDeployments. Is it better to keep it `*vmoprv1.AffinitySpec` for less Mem usage?

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+				Reason: infrav1.VSphereMachineVirtualMachineWaitingForVirtualMachineGroupV1Beta2Reason,
+			})
+			log.V(4).Info("Waiting for VirtualMachineGroup membership, requeueing", "VM Name", supervisorMachineCtx.Machine.Name)
+			return true, nil
```
Sure, filed issue https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/3672

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+// - Placement Persistence: Persisting the MachineDeployment-to-Zone mapping (placement decision) as a
+// metadata annotation on the VMG object. This decision is crucial for guiding newer VMs created
+// during Day-2 operations such as scaling, upgrades, and remediations, ensuring consistency. This is also due to
+// a known limitation of underlying services.
```
Fixed.

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+			cluster.Name, cluster.Namespace)
+	}
+
+	desiredVMG := &vmoprv1.VirtualMachineGroup{}
```
Updated to use `vmg`

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+
+		// Calculate expected Machines of all MachineDeployments.
+		// CAPV retrieves placement decisions from the VirtualMachineGroup to guide
+		// day-2 VM placement. At least one VM is required for each MachineDeployment.
```
Since this check only happens when VMG is not created yet, so for day-2 scale from zero, it won't go into this case.
In addition for day-1, we'll block create Cluster with MD.replica == 0.

zhanggbj replied:
For day-2, since we'll get vmg successfully in line , so we won't run into this line.

sbueringer commented on controllers/vmware/virtualmachinegroup_controller.go:
```diff
@@ -0,0 +1,109 @@
+// VSphereMachineToCluster maps VSphereMachine events to Cluster reconcile requests.
+// This handler only processes VSphereMachine objects for Day-2 operations when VMG could be found, ensuring
+// VMG member list in sync with VSphereMachines. If no corresponding VMG is found, this is a no-op.
+func (r *VirtualMachineGroupReconciler) VSphereMachineToCluster(ctx context.Context, a ctrlclient.Object) []reconcile.Request {
```
Yes!

zhanggbj commented on config/rbac/role.yaml:
```diff
@@ -264,6 +265,12 @@ rules:
+- apiGroups:
+  - vmoperator.vmware.com
+  resources:
+  - virtualmachinegroups/status
```
Filed an issue for tracking: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/3681

zhanggbj commented:
@sbueringer @fabriziopandini 
I've addressed above comments, please help to take another review, thanks!

BTW, I'm still improving logging and godoc incrementally, feel free to point out if anything missing.

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +199,120 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+		}
+		mdNames := []string{}
+		for _, machineDeployment := range machineDeployments.Items {
+			if machineDeployment.Spec.Template.Spec.FailureDomain == "" && machineDeployment.Name != mdName {
```
is it correct to not consider machine deployments without FailureDomain?

As a user, I would like that anti affinity considers all the MachineDeployments, not only a subset (only exception should be the MachineDeployment the machine belongs to)

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+
+		affInfo.affinitySpec = &vmoprv1.AffinitySpec{
+			VMAffinity: &vmoprv1.VMAffinitySpec{
+				RequiredDuringSchedulingPreferredDuringExecution: []vmoprv1.VMAffinityTerm{
```
This comment seems not yet addressed

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+			},
+		}
+		if len(mdNames) > 0 {
+			affInfo.affinitySpec.VMAntiAffinity.PreferredDuringSchedulingPreferredDuringExecution = append(
```
This comment seems not yet addressed

fabriziopandini commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,392 @@
+// - Batch Coordination: Gating the initial creation of the VMG until all expected worker
+// VSphereMachines are present. It will compare the desired count of VirtualMachines
+// derived from the sum of all desired MachineDeployment Replicas) against the count
+// of existing, matching VSphereMachine objects.
```
I think this definition better reflects what the code is doing
```suggestion
// - Batch Coordination: Gating the initial creation of the VMG until for the first time all the  
//    MachineDeployment replicas will have a corresponding VSphereMachine.
//    Once this condition is met, the VirtualMachineGroup is created considering
//    the initial set of machines for the initial placement decision.
//    When the VirtualMachineGroup reports the placement decision, then finally
//    creation of VirtualMachines is unblocked.
```

fabriziopandini commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,392 @@
+		return reconcile.Result{}, err
+	}
+
+	log = log.WithValues("VirtualMachineGroup", klog.KObj(cluster))
```
```suggestion
	// Note: VirtualMachineGroup is going to have same name and namespace of the cluster.
	// Using cluster here, because VirtualMachineGroup is created only after initial placement completes.
	log = log.WithValues("VirtualMachineGroup", klog.KObj(cluster))
```

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+		Name:      cluster.Name,
+	}
+
+	if err := r.Client.Get(ctx, *key, desiredVMG); err != nil {
```
Added comments.

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+	current, err := getCurrentVSphereMachines(ctx, r.Client, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to get current VSphereMachine of cluster %s/%s",
+			cluster.Name, cluster.Namespace)
```
Refined logging inside getCurrentVSphereMachines.

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+
+		// Wait for all intended VSphereMachines corresponding to MachineDeployment to exist only during initial Cluster creation.
+		current := int32(len(current))
+		if current < expected {
```
Fixed.

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+		if current < expected {
+			log.Info("current VSphereMachines do not match expected", "Expected:", expected,
+				"Current:", current, "ClusterName", cluster.Name, "Namespace", cluster.Namespace)
+			return reconcile.Result{RequeueAfter: reconciliationDelay}, nil
```
Removed RequeueAfter.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +199,120 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+			client.MatchingLabels{clusterv1.ClusterNameLabel: supervisorMachineCtx.Cluster.Name}); err != nil {
+			return false, err
+		}
+		mdNames := []string{}
```
Fixed.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -494,6 +619,21 @@ func (v *VmopMachineService) reconcileVMOperatorVM(ctx context.Context, supervis
+		// initial cluster creation (Day 1) to determine the VM's initial location.
+		// These rules DO NOT impact new VMs created later during scaling (Day 2).
+		// Day 2 placement often relies on dynamic information derived from
+		// VirtualMachineGroup annotations.
```
Fixed.

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,392 @@
+		// Wait for all intended VSphereMachines corresponding to MachineDeployment to exist only during initial Cluster creation.
+		currentVSphereMachineCount := int32(len(currentVSphereMachines))
+		if currentVSphereMachineCount < expectedVSphereMachineCount {
+			log.Info("Waiting for expected VSphereMachines to be created", "Expected:", expectedVSphereMachineCount,
```
Updated.

zhanggbj commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,400 @@
+
+		// Check if this VM belongs to any of our target MachineDeployments.
+		// Annotation format is "zone.cluster.x-k8s.io/{machine-deployment-name}".
+		for _, md := range machineDeployments {
```
Make sense. Fixed it.

zhanggbj reviewed (commented):
All comments are addressed, including the ones missed by mistake last time.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+		}
+	}
+
+	if supervisorMachineCtx.Machine.Spec.FailureDomain != "" {
```
Comments are added.


> Q: shouldn't we set the failure domain on the VM also when the failure domain is empty but the Machine deployment has been already placed
> 
> @sbueringer I don't remember if this was discussed, do you?

I didn't quite get this one. For a Machine without failureDomain, it should be using auto placement, so VM should be created and placed using affinity rules, or VM should be recreated based on the placement annotation in day-2.

sbueringer replied:
Discussed with Fabrizio, current code here is fine

zhanggbj commented:
/test pull-cluster-api-provider-vsphere-e2e-vcsim-supervisor-main

sbueringer commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
 	supervisorMachineCtx.VSphereMachine.Status.VMStatus = vmwarev1.VirtualMachineStatePending
 
+	var affInfo affinityInfo
+	if feature.Gates.Enabled(feature.NodeAutoPlacement) &&
```
I have no idea if standalone Machines / MachineSets are supported by CAPV today. E.g. CAPV govmomi always assumes that a worker Machine is part of a MD for cluster-modules (afaik).
No idea about supervisor mode, requires some research.

zhanggbj commented:
Pushed a small change and checking UT failures.

fabriziopandini commented on controllers/vmware/virtualmachinegroup_reconciler.go:
```diff
@@ -0,0 +1,488 @@
+				continue
+			}
+			if member.Placement != nil && member.Placement.Zone != "" {
+				log.Info(fmt.Sprintf("MachineDeployment %s has been placed to failure domanin %s", md.Name, member.Placement.Zone), "MachineDeployment", klog.KObj(&md))
```
nit
```suggestion
				log.Info(fmt.Sprintf("MachineDeployment %s has been placed to failure domain %s", md.Name, member.Placement.Zone), "MachineDeployment", klog.KObj(&md))
```

fabriziopandini commented on controllers/vmware/virtualmachinegroup_reconciler_test.go:
```diff
@@ -0,0 +1,1227 @@
+		want            bool
+	}{
+		{
+			name: "Should create a VMG if all the expected VSphereMachines exists",
```
Let's add a unit test ensuring that a deleting vSphereMachine blocks.
I would suggest to use this test case as a setup and add an additional deleting machine

zhanggbj replied:
@fabriziopandini 
Trying to understand this scenario, does it mean a deleting vSphereMachine won't be added to the VMGroup member list?

If yes, this case should cover it `name: "Should create a VMG if all the expected VSphereMachines exists, deleting MD should be ignored"`

fabriziopandini replied:
This is what is in the current implementation, and it makes sense to me because it makes sure that we start a placement decision only when the list of vSphereMachines is stable.

Let's please add a unit test and do not change the code

fabriziopandini commented on controllers/vmware/virtualmachinegroup_reconciler_test.go:
```diff
@@ -0,0 +1,1227 @@
+						clusterv1.ClusterNameLabel: cluster.Name,
+					},
+					Annotations: map[string]string{
+						fmt.Sprintf("%s/%s", vmoperator.ZoneAnnotationPrefix, "md3"): "zone1", // failureDomain for md3 is explicitly set by the user
```
Let's add an annotation without the ZoneAnnotationPrefix and check it is preserved

zhanggbj replied:
Updated.

fabriziopandini commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,27 +199,161 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+			client.MatchingLabels{clusterv1.ClusterNameLabel: supervisorMachineCtx.Cluster.Name}); err != nil {
+			return false, err
+		}
+		othermMDNames := []string{}
```
nit
```suggestion
		otherMDNames := []string{}
```

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,27 +199,161 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+				Reason:  infrav1.VSphereMachineVirtualMachineWaitingForVirtualMachineGroupV1Beta2Reason,
+				Message: fmt.Sprintf("Waiting for VSphereMachine's VirtualMachineGroup %s to exist", key),
+			})
+			log.V(4).Info(fmt.Sprintf("Waiting for VirtualMachineGroup %s, requeueing", key.Name), "VirtualMachineGroup", klog.KRef(key.Namespace, key.Name))
```
Updated.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,27 +199,161 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+				Type:    infrav1.VSphereMachineVirtualMachineProvisionedV1Beta2Condition,
+				Status:  metav1.ConditionFalse,
+				Reason:  infrav1.VSphereMachineVirtualMachineWaitingForVirtualMachineGroupV1Beta2Reason,
+				Message: fmt.Sprintf("Waiting for VirtualMachineGroup %s membership", klog.KRef(key.Namespace, key.Name)),
```
Fixed.

zhanggbj commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -188,27 +199,161 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
+				Reason:  infrav1.VSphereMachineVirtualMachineWaitingForVirtualMachineGroupV1Beta2Reason,
+				Message: fmt.Sprintf("Waiting for VirtualMachineGroup %s membership", klog.KRef(key.Namespace, key.Name)),
+			})
+			log.V(4).Info(fmt.Sprintf("Waiting for VirtualMachineGroup %s membership, requeueing", key.Name), "VirtualMachineGroup", klog.KRef(key.Namespace, key.Name))
```
Fixed.

zhanggbj commented:
`pull-cluster-api-provider-vsphere-e2e-supervisor-blocking-main` failed due to infra failure and I'm not able to re-trigger again. Need to retry later.
